collect_app/src/main/java/org/odk/collect/android/instancemanagement/InstanceAutoSender.kt

```diff
@@ -0,0 +1,97 @@
+package org.odk.collect.android.instancemanagement
+
+import android.content.Context
+import org.odk.collect.analytics.Analytics
+import org.odk.collect.android.R
+import org.odk.collect.android.formmanagement.InstancesAppState
+import org.odk.collect.android.gdrive.GoogleAccountsManager
+import org.odk.collect.android.gdrive.GoogleApiProvider
+import org.odk.collect.android.notifications.Notifier
+import org.odk.collect.android.upload.FormUploadException
+import org.odk.collect.android.utilities.ChangeLockProvider
+import org.odk.collect.android.utilities.FormsRepositoryProvider
+import org.odk.collect.android.utilities.InstanceUploaderUtils
+import org.odk.collect.android.utilities.InstancesRepositoryProvider
+import org.odk.collect.forms.FormsRepository
+import org.odk.collect.forms.instances.Instance
+import org.odk.collect.forms.instances.InstancesRepository
+import org.odk.collect.permissions.PermissionsProvider
+import org.odk.collect.settings.SettingsProvider
+import org.odk.collect.settings.keys.ProjectKeys
+import org.odk.collect.shared.settings.Settings
+
+class InstanceAutoSender(
+    private val context: Context,
+    private val changeLockProvider: ChangeLockProvider,
+    private val notifier: Notifier,
+    private val analytics: Analytics,
+    private val formsRepositoryProvider: FormsRepositoryProvider,
+    private val instancesRepositoryProvider: InstancesRepositoryProvider,
+    private val googleAccountsManager: GoogleAccountsManager,
+    private val googleApiProvider: GoogleApiProvider,
+    private val permissionsProvider: PermissionsProvider,
+    private val settingsProvider: SettingsProvider,
+    private val instancesAppState: InstancesAppState
+) {
+    fun autoSendInstances(projectId: String): Boolean {
+        val formsRepository = formsRepositoryProvider.get(projectId)
```

Comment:
I was thinking about it but it would basically moving dependencies from one class to another. I can't construct the whole `ProjectSandbox` using dagger and inject to `AutoSendTaskSpec`  because I don't know what project (`projectId`) it should represent. That means I would need to construct it in `AutoSendTaskSpec` and to do that I would need to inject all the dependencies it needs there instead of to `InstanceAutoSender`. I don't think it makes a lot of sense.... maybe if it would make testing `InstanceAutoSender` easier but it's related to the question above.

Reply:
Oh I still think you'd want to use it in the same way as `FormsUpdater` rather than injecting it with Dagger - constructing it based on the `projectId`. I was just thinking of it as a way to clean up the noise from grabbing all the repositories etc for the Project. It's just an aesthetic change really so you don't have a big preamble before the logic.